File: epi_judge_java/epi/IsValidSudoku.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class IsValidSudoku {

    @EpiTest(testDataFile = "is_valid_sudoku.tsv")
    // Check if a partially filled matrix has any conflicts.
    public static boolean isValidSudoku(List<List<Integer>> partialAssignment) {
        Set<String> set = new HashSet<>();
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                int value = partialAssignment.get(r).get(c);
                if (value > 0) {
                    if (!set.add("Seen " + value + " on " + r + " row")
                            || !set.add("Seen " + value + " on " + c + " column")
                            || !set.add("Seen " + value + " on " + (r / 3 * 3 + c / 3) + " subgrid")) {

                        return false;
                    }
                }
            }
        }

        return true;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IsValidSudoku.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/is_valid_sudoku.cc
#include <vector>

#include "test_framework/generic_test.h"
using namespace std;
// Check if a partially filled matrix has any conflicts.
bool IsValidSudoku(const vector<vector<int> > &partial_assignment) {
    unordered_set<string> checker;
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            if (partial_assignment[i][j] == 0) continue;
            string row = "row " + to_string(i) + " contains " + to_string(partial_assignment[i][j]);
            string column = "column " + to_string(j) + " contains " + to_string(partial_assignment[i][j]);
            string cell = "cell " + to_string(i / 3) + ":" + to_string(j / 3) + " contains " + to_string(partial_assignment[i][j]);
            if (checker.count(row) || checker.count(column) || checker.count(cell)) return false;
            checker.insert(row);
            checker.insert(column);
            checker.insert(cell);
        }
    }

    return true;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"partial_assignment"};
    return GenericTestMain(args, "is_valid_sudoku.cc", "is_valid_sudoku.tsv",
                           &IsValidSudoku, DefaultComparator{}, param_names);
}
